### PR TITLE
Test dynamo wrapper

### PR DESCRIPTION
**Pull request by mohit**
A prototype for get/set results to DyanamoDB. Supports some functionality of `gearcmd`.

No tests yet since not currently sure if this is the model we will be taking.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,15 @@
...
-Utility to manage inputs and outputes of batch worker executions
+
+Utility to manage inputs and outputes of batch worker executions. Uses DynamoDB as backend
+
+Owned by eng-infra
+
+## Usage
+
+```
+gearcmd -name <job id> -input <key-id> -cmd <command>
```
Comment (rgarcia) [inline README.md]: s/gearcmd/batchcli/

File: runner/batch.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+	"os"
+	"strings"
+)
+
+type BatchJob struct {
+	JobId              string
+	Queue              string
+	ComputeEnvironment string
+	DependencyIds      []string
+}
```
Comment (rgarcia) [inline runner/batch.go]: without a good understanding of AWS Batch, it's hard to get oriented on what this struct represents. Comments on this struct + field values would make it easier!

File: runner/batch.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+// NewBatchJobFromEnv returns a new BatchJob by reading
+// values from the envrionment
```
Comment (rgarcia) [inline runner/batch.go]: s/envrionment/environment/

File: runner/batch.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+	jobId := os.Getenv("AWS_BATCH_JOB_ID")
+	queue := os.Getenv("AWS_BATCH_JQ_NAME")
+	computeEnv := os.Getenv("AWS_BATCH_CE_NAME")
+
+	if jobId == "" || queue == "" || computeEnv == "" {
+		return BatchJob{},
+			errors.New("AWS Batch environment variables not found")
+	}
+
+	depsStr := os.Getenv("_BATCH_DEPEDENCIES")
```
Comment (rgarcia) [inline runner/batch.go]: comment about why this environment variable is different from the others

File: runner/runner.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+		"inputs": t.inputs,
+		"cmd":    t.cmd,
+		"job-id": t.job.JobId,
+	})
+
+	cmd := exec.Command(t.cmd, t.inputs...)
+	cmd.Env = os.Environ()
+
+	// Write the stdout and stderr of the process to both this process' stdout and stderr
+	// and also write it to a byte buffer so that we can return it with the Gearman job
```
Comment (rgarcia) [inline runner/runner.go]: mentions Gearman

File: runner/runner.go [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+	results, err := store.GetResults(job.DependencyIds)
+	if err != nil {
+		return TaskRunner{}, err
+	}
+	// postfix the results of previous jobs on the cmd passed through
+	// the CLI
+	// example:
+	// 		batchcli -cmd echo hello there
+	// 		results = [{"json":"true"}, {}]
+	//      exec(echo, ["hello", "there", '{"json":"true"}', '{}'])
```
Comment (rgarcia) [inline runner/runner.go:61]: interesting approach to getting the results of dependent jobs... Is there a way for the user to know which positional argument belongs to which dependency, or will they have to have an assumption around dependency ordering?

File: runner/store.go [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+	return outputs, nil
+}
+
+func NewDynamoStore(table string) (ResultsStore, error) {
+	sess, err := session.NewSession(&aws.Config{Region: aws.String("us-east-1")})
+	if err != nil {
+		return DynamoStore{}, err
+	}
+
+	client := dynamodb.New(sess)
```
Comment (rgarcia) [inline runner/store.go]: IMO constructor here should take in [`dynamodbiface.DynamoDBAPI`](https://godoc.org/github.com/aws/aws-sdk-go/service/dynamodb/dynamodbiface#DynamoDBAPI) and the initialization of the session and client object should happen in callers of `NewDynamoStore()`.

File: runner/store.go [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+func (d DynamoStore) writeResult(success bool, key string, result string) error {
+	_, err := d.client.PutItem(&dynamodb.PutItemInput{
+		Item: map[string]*dynamodb.AttributeValue{
+			"Key": {
+				S: aws.String(key),
+			},
+			"Result": {
+				S: aws.String(result),
```
Comment (rgarcia) [inline runner/store.go]: probably want to put a check here for limits around key/val and total item size in dynamo, since this is the raw stdout or stderr

**Review by rgarcia — COMMENTED** [review]
did an initial pass, looks nice. Would be nice to comment on the Batch-specific stuff in here. Besides that the storing results in dynamo code was pretty easy to follow.

Will this eventually interface with the Batch API in some way, or is its sole purpose capturing / storing results in a standard way?

**Comment by mohit** [comment]
Thanks. Exactly the kinds of comments that I was looking for. This should ideally not interact with the AWS Batch API. It's only a wrapper to allow for I/O capture and for passing results between jobs.

File: runner/runner.go (hunk 6 above)
Comment (mohit) [reply runner/runner.go:61]: Not sure yet. I've been wondering if I should pass in the dependent job-name (which I don't have access to without additional env vars) or just the dependency job-ids.

One other direction I had considered is to write the job-name to dyanmo and then pull out results as a map:

```bash
batchcli -cmd echo hello there
results = {"job-one-name": {"json":"true"}, "job-two-name": {}}

exec(echo, ["hello", "there", {"job-one-name": {"json":"true"}, "job-two-name": {}}]
```
This is simpler, but now it means that there is a dependency between job-names in the workflow manager and the Worker code.  

Another option would be to just pass the results as a list `[]` and the underlying Worker should figure out which object is from what dependency.

**Comment by mohit** [comment]
@rgarcia thanks again for the feedback. Added comments and fixed references I had copied over from gearcmd. Still not sure about how to best pass outputs of dependencies over command line but I'm hoping to figure that out over the rest of this week. https://github.com/Clever/batchcli/pull/1/commits/76c2adc9f45b8190afd89776322f80d3a8690044

**Pull request merged**